src/passes/TypeMerging.cpp

```diff
@@ -242,27 +242,34 @@ void TypeMerging::run(Module* module_) {
   // Merging can unlock more sibling merging opportunities because two identical
   // types cannot be merged until their respective identical parents have been
   // merged in a previous step, making them siblings.
-  //
-  // If we merge siblings, we also need to refinalize because the LUB of merged
-  // siblings is the merged type rather than their common supertype after the
-  // merge.
-  bool refinalize = false;
   merge(Supertypes);
   for (int i = 0; i < MAX_ITERATIONS; ++i) {
-    if (!merge(Siblings)) {
-      break;
-    }
-    refinalize = true;
+    merge(Siblings);
   }
 
   applyMerges();
 
-  if (refinalize) {
-    ReFinalize().run(getPassRunner(), module);
-  }
+  // If we merge siblings, we also need to refinalize because the LUB of merged
+  // siblings is the merged type rather than their common supertype after the
+  // merge. This can happen in merge(Siblings), but also in merge(Supertypes),
+  // since we may end up merging B1 to its super A, and also B2 to the same
+  // super A, ending up with B1 and B2 now equal - in that case the siblings are
+  // now both equal (to the parent), allowing an exact LUB:
+  //
+  //  (select (result A))
+  //   (B1)
+  //   (B2)
+  //  )
+  // =>
+  //  (select (result (exact A))) ;; result is now exact
+  //   (A) ;; both are
+  //   (A) ;; now A
+  //  )
+  //
+  ReFinalize().run(getPassRunner(), module);
```

Comment:
We could still check whether any merging occurred before refinalizing.

Reply:
Good idea, done.